Jenkinsfile

```diff
@@ -3,34 +3,55 @@ properties([
     durabilityHint('PERFORMANCE_OPTIMIZED'),
     buildDiscarder(logRotator(numToKeepStr: '5')),
 ])
-parallel kind: {
-    node('docker') {
-        timeout(90) {
-            checkout scm
-            withEnv(["WSTMP=${pwd tmp: true}"]) {
-                try {
-                    sh 'bash kind.sh'
-                    dir (WSTMP) {
-                        junit 'surefire-reports/*.xml'
-                    }
-                } finally {
-                    dir (WSTMP) {
-                        if (fileExists('kindlogs/docker-info.txt')) {
-                            archiveArtifacts 'kindlogs/'
+
+def splits
+stage('Determine splits') {
+    node('linux') {
+        checkout scm
+        splits = splitTests parallelism: count(2), generateInclusions: true, estimateTestsFromFiles: true
+    }
+}
+stage('Tests') {
+    def branches = [:]
+    branches['failFast'] = true
+
+    for (int i = 0; i < splits.size(); i++) {
+        def num = i
+        def split = splits[num]
+        def index = num + 1
+        branches["kind-${index}"] = {
+            node('docker') {
+                timeout(90) {
+                    checkout scm
+                    try {
+                        writeFile file: (split.includes ? "$WORKSPACE_TMP/includes.txt" : "$WORKSPACE_TMP/excludes.txt"), text: split.list.join("\n")
+                        writeFile file: (split.includes ? "$WORKSPACE_TMP/excludes.txt" : "$WORKSPACE_TMP/includes.txt"), text: ''
+                        sh './kind.sh -Dsurefire.includesFile="$WORKSPACE_TMP/includes.txt" -Dsurefire.excludesFile="$WORKSPACE_TMP/excludes.txt"'
```

Comment:
Hmm, so this winds up installing Kind twice per build. Expensive, but I guess worth the reduction in clock time.

Reply:
yeah the whole `kind` setup is ~2 min I think.

Reply:
Including downloading images? Not bad.

Reply:
We could include `kind` in the builder base image. WDYT?

Reply:
Well the thing is that the `kind` executable is just a few Mb; most of the download bandwidth is for the “Kindest node” images when creating a cluster.

Anyway we pin a specific version in plugin sources https://github.com/jenkinsci/kubernetes-plugin/blob/58918882cfb7096388049f60b4e269a8bfd5e10f/kind.sh#L7 and do really need to control exactly when that is updated (as part of a PR which may be adding or removing support for some K8s versions), so a binary baked into the VM would be useless.

Reply:
Thanks for the detailed response @jglick